src/compute/sort/row/mod.rs

```diff
@@ -0,0 +1,1041 @@
+    // as identifying the offsets of the written rows
+    let mut cur_offset = 0_usize;
+    for l in lengths {
+        offsets.push(cur_offset);
+        cur_offset = cur_offset.checked_add(l).expect("overflow");
+    }
+
+    let buffer = vec![0_u8; cur_offset];
+
+    Ok(Rows {
+        buffer: buffer.into(),
+        offsets: offsets.into(),
+        fields,
+    })
+}
+
+/// Encodes a column to the provided [`Rows`] incrementing the offsets as it progresses
+fn encode_column(
+    out: &mut Rows,
+    column: &Box<dyn Array>,
+    opts: SortOptions,
+    dictionary: Option<&[Option<&[u8]>]>,
+) {
+    match column.data_type().to_physical_type() {
+        PhysicalType::Primitive(primitive) => {
+            with_match_primitive_without_interval_type!(primitive, |$T| {
+                let column = column
+                    .as_any()
+                    .downcast_ref::<PrimitiveArray<$T>>()
+                    .unwrap()
+                    .iter()
+                    .map(|v| v.map(|v| *v));
+                fixed::encode(out, column, opts);
+            })
+        }
+        PhysicalType::Null => {}
+        PhysicalType::Boolean => fixed::encode(
+            out,
+            column.as_any().downcast_ref::<BooleanArray>().unwrap(),
+            opts,
+        ),
+        PhysicalType::Binary => {
+            variable::encode(
+                out,
+                column
+                    .as_any()
+                    .downcast_ref::<BinaryArray<i32>>()
+                    .unwrap()
+                    .iter(),
+                opts,
+            );
+        }
+        PhysicalType::LargeBinary => {
+            variable::encode(
+                out,
+                column
+                    .as_any()
+                    .downcast_ref::<BinaryArray<i64>>()
+                    .unwrap()
+                    .iter(),
+                opts,
+            );
+        }
+        PhysicalType::Utf8 => variable::encode(
+            out,
+            column
+                .as_any()
+                .downcast_ref::<Utf8Array<i32>>()
+                .unwrap()
+                .iter()
+                .map(|x| x.map(|x| x.as_bytes())),
+            opts,
+        ),
+        PhysicalType::LargeUtf8 => variable::encode(
+            out,
+            column
+                .as_any()
+                .downcast_ref::<Utf8Array<i64>>()
+                .unwrap()
+                .iter()
+                .map(|x| x.map(|x| x.as_bytes())),
+            opts,
+        ),
+        PhysicalType::Dictionary(k) => match_integer_type!(k, |$T| {
+            let column = column
+                .as_any()
+                .downcast_ref::<DictionaryArray<$T>>()
+                .unwrap();
+            encode_dictionary(out, column, dictionary.unwrap(), opts);
+        }),
+        t => unimplemented!("not yet implemented: {:?}", t),
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
move to `tests/` (all tests to public APIs go there so that we verify that the functions and structs are available from the crate instead of `super`)

Reply:
How about some tests using private methods of a module?

Reply:
Should we spilit the two types of tests?

Reply:
exactly - tests to public functionality go to public, rest stays here :)